test/workspace.hpp

```diff
@@ -0,0 +1,160 @@
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#pragma once
+
+#include <hip/hip_runtime.h>
+
+#include "get_handle.hpp"
+
+class Workspace
+{
+
+    // RAII class for hip allocations
+    class GPUBuffer
+    {
+    public:
+        GPUBuffer() = default;
+
+        explicit GPUBuffer(size_t num_bytes)
+        {
+            assert(num_bytes > 0);
+            auto s = hipMalloc(&buf_, num_bytes);
+            if(s != hipSuccess || buf_ == nullptr)
+            {
+                std::abort();
+            }
+        }
+
+        ~GPUBuffer()
+        {
+            auto s = hipFree(buf_);
+            buf_   = nullptr;
+            if(s != hipSuccess)
+            {
+                std::abort();
+            }
+        }
+
+        void* ptr() { return buf_; }
+        void* ptr() const { return buf_; }
+
+        GPUBuffer(const GPUBuffer&) = delete;
+        GPUBuffer& operator=(const GPUBuffer&) = delete;
+
+        GPUBuffer(GPUBuffer&& that) noexcept : buf_(that.buf_)
+        {
+            that.buf_ = nullptr; // take over ownership
+        }
+
+        GPUBuffer& operator=(GPUBuffer&& that) noexcept
+        {
+            std::swap(this->buf_, that.buf_);
+            return *this;
+        }
+
+    private:
+        void* buf_ = nullptr;
+    };
+
+public:
+    Workspace() = default;
+
+    Workspace(const Workspace&) = delete;
+    Workspace& operator=(const Workspace&) = delete;
+    Workspace(Workspace&&)                 = default;
+    Workspace& operator=(Workspace&&) = default;
+
+    size_t size() const { return sz_; }
+
+    void resize(size_t sz_in_bytes)
+    {
+        sz_ = sz_in_bytes;
+        AdjustToSize();
+    }
+
+    // for use in miopen .*GetWorkSpaceSize() methods where a pointer to size_t is
+    // passed to capture the size. Must call AdjustToSize() after calling such a method
+    size_t* SizePtr() { return &sz_; }
```

Comment:
Since this class is for the testing purposes, that sort of resize can be checked and reported.
Using `sz_` and `old_sz_` may help to detect unexpected `sz_` update.
Both members can be synchronized on `AdjustToSize` and checked to be equal on each other operation.